### PR TITLE
DDP-6127 - implement single participant lookup by a participant guid

### PR DESCRIPTION
**Pull request by evgeniipr**
## Context

Lookup a participant by a guid. 
Implemented a route `ParticipantLookupByGuidRoute` which handles a request handles request `GET /studies/{studyGuid}/participants/{userGuid}`. It utilizes a service `ParticipantsLookupService` which used for participants lookup by a query string (full-text search). The same implementation of this service is used: `ESParticipantsLookupService`. This implementation makes a search in Pepper ElasticSearch database. 
In order to distinguish types of search it is defined 2 types: FULL_TEXT_SEARCH_BY_QUERY_STRING, BY_PARTICIPANT_GUID.

## Checklist

- [x] I have labeled the type of changes involved using the `C-*` labels.
- [x] I have assessed potential risks and labeled using the `R-*` labels.
- [x] I have considered error handling and alerts, and added `L-*` labels as needed.
- [x] I have considered security and privacy, and added `I-*` labels as needed
- [x] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.
- [x] If applicable, I have discussed the analytics needs at both a platform and study level with Product and instrumented code accordingly.
- [x] If applicable, my UI/UX changes have passed muster with Product/Design via an over-the-shoulder review, screenshots, etc.

_If unsure or need help with any of the above items, add the `help wanted` label. For items that starts with `If applicable`, if it is not applicable, check it off and add `n/a` in front._

## FUD Score

_Overall, how are you feeling about these changes?_

- [ ] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [x] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [ ] I have written automated positive tests
- [ ] I have written automated negative tests
- [x] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [x] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

File: pepper-apis/docs/specification/src/pepper.yml [hunk 1]
```diff
@@ -66,6 +66,8 @@ paths:
     $ref: 'endpoints/admin.studies.participants.yml'
   /admin/studies/{study}/participants-lookup:
     $ref: 'endpoints/admin.studies.participants-lookup.yml'
+  /admin/studies/{study}/participants/{user}:
+    $ref: 'endpoints/admin.studies.participant-lookup-by-guid.yml'
```
Comment (yufengwng) [inline pepper-apis/docs/specification/src/pepper.yml]: Can you rename the file to `admin.studies.participant.yml` and move this block to be right below `/admin/studies/{study}/participants`?

File: pepper-apis/docs/specification/src/endpoints/admin.studies.participant-lookup-by-guid.yml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  description: |
+    Lookup a participant by participant guid (user guid).
+
+    It should return a single record with a participant info.
+    If nothing is found then error 404 is returned.
+
+    Note: requires study admin ID token.
+  parameters:
+    - $ref: '../pepper.yml#/components/parameters/userGuid'
+    - $ref: '../pepper.yml#/components/parameters/studyGuid'
```
Comment (yufengwng) [inline pepper-apis/docs/specification/src/endpoints/admin.studies.participant-lookup-by-guid.yml]: I think it's better to sort this in the same order as how the parameters show up in the path, so it should be `studyGuid` then `userGuid`.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/DataDonationPlatform.java [hunk 3]
```diff
@@ -371,6 +372,8 @@ private static void start() throws MalformedURLException {
         post(API.ADMIN_STUDY_USER_LOGIN_ACCOUNT, new AdminCreateUserLoginAccountRoute(), jsonSerializer);
         post(API.ADMIN_STUDY_PARTICIPANTS_LOOKUP,
                 new ParticipantsLookupRoute(new ESParticipantsLookupService(esClient)), responseSerializer);
+        get(API.ADMIN_STUDY_PARTICIPANT_LOOKUP_BY_GUID,
+                new ParticipantLookupByGuidRoute(new ESParticipantsLookupService(esClient)), responseSerializer);
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/DataDonationPlatform.java]: Can we rename the class to `AdminParticipantLookupByGuidRoute`? Let's add `Admin` prefix to `ParticipantsLookupRoute` too.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/util/RouteUtil.java [hunk 4]
```diff
@@ -223,6 +228,20 @@ public static void haltIfError(User user, String participantGuid, String studyGu
         }
     }
 
+    public static void haltError(int status, String code, String msg) {
+        LOG.warn(msg);
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/util/RouteUtil.java]: One drawback of abstracting out a method like this is that this `LOG` object now refers to the `RouteUtil` class and not `AdminParticipantsLookupRoute`. So you won't be able to tell where the log entry came from in the logs.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/route/ParticipantLookupByGuidRoute.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
+package org.broadinstitute.ddp.route;
+
+import static org.apache.http.HttpStatus.SC_NOT_FOUND;
+import static org.broadinstitute.ddp.constants.ErrorCodes.USER_NOT_FOUND;
+import static org.broadinstitute.ddp.route.ParticipantsLookupRoute.handleParticipantLookupException;
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/route/ParticipantLookupByGuidRoute.java]: Feels a little weird to static import something from a route...

File: pepper-apis/docs/specification/src/pepper.yml (hunk 1 above)
Comment (evgeniipr) [reply pepper-apis/docs/specification/src/pepper.yml]: Fixed

File: pepper-apis/docs/specification/src/endpoints/admin.studies.participant-lookup-by-guid.yml (hunk 2 above)
Comment (evgeniipr) [reply pepper-apis/docs/specification/src/endpoints/admin.studies.participant-lookup-by-guid.yml]: Fixed

File: pepper-apis/src/main/java/org/broadinstitute/ddp/DataDonationPlatform.java (hunk 3 above)
Comment (evgeniipr) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/DataDonationPlatform.java]: Fixed

File: pepper-apis/src/main/java/org/broadinstitute/ddp/util/RouteUtil.java (hunk 4 above)
Comment (evgeniipr) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/util/RouteUtil.java]: Fixed this: added method 'haltError' to a Route-class.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/route/ParticipantLookupByGuidRoute.java (hunk 5 above)
Comment (evgeniipr) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/route/ParticipantLookupByGuidRoute.java]: Fixed this: moved common code to a separate util class.

**Review by yufengwng — APPROVED** [review]

**Pull request merged**